lib/process_balancer/manager.rb

```diff
@@ -117,11 +117,26 @@ def update_state
       [exists, msg]
     end
 
+    def stale_workers(connection, workers)
+      workers.reject { |i| connection.exists?(i) }
+    end
+
     def update_process_index
       redis do |c|
         c.watch(PROCESSES_KEY)
 
-        workers     = c.lrange(PROCESSES_KEY, 0, -1)
+        workers   = c.lrange(PROCESSES_KEY, 0, -1)
+        stale_ids = workers.reject { |i| c.exists?(i) }
```

Comment:
what if we move this to its own method? I think it will be much more clear about what the intention of `update_process_index` is, maybe something like:

It looks like there are 3 or 4 actions that we need to perform

```ruby
def update_process_index
  get_workers(connection)
  remove_stale_workers(connection, workers)
  add_instance(identity) if index.nil?
  
  # Now we should have the right process_index and process_count
end
```

what do you think?

Reply:
The fun part of it is that I need it to be in a "transaction" within Redis.. As I originally started doing it like that and it became nasty with all the cross-communication between the pieces needed.  (that c.watch is part of the transaction handling)